nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/handler/LogRecordBuffer.java

```diff
@@ -128,36 +199,60 @@ public void add(final GlassFishLogRecord record) {
      */
     private void addWithTimeout(final GlassFishLogRecord record) {
         try {
-            if (this.pendingRecords.offer(record)) {
-                return;
-            }
-            if (this.pendingRecords.offer(record, this.maxWait, TimeUnit.SECONDS)) {
+            if (availableCapacity.tryAcquire(maxWait, TimeUnit.SECONDS)) {
+                pendingRecords.add(record);
                 return;
             }
         } catch (final InterruptedException e) {
             // do nothing
         }
 
-        this.pendingRecords.clear();
-        // note: the record is not meaningful for the message. The cause is in another place.
-        this.pendingRecords.offer(new GlassFishLogRecord(Level.SEVERE, //
-            this + ": The buffer was forcibly cleared after " + maxWait + " s timeout for adding another log record." //
-                + " Log records were lost." //
-                + " It might be caused by a recursive deadlock," //
-                + " you can increase the capacity or the timeout to avoid this.", false));
+        lock.lock();
+        try {
+            try {
+                if (availableCapacity.tryAcquire(0, TimeUnit.SECONDS)) {
```

Comment:
Why this attempt?

Reply:
> Why this attempt?

Anothe thread may have already drops all waiting records and we have space in the buffer. Some minor optimization.